### PR TITLE
Test updateWith() state machine

### LINKED ISSUE #7288: Test PaymentRequest .updateWith()
Needed tests: 
  * [x] Follow the spec to successful update.
  * [x] Abort the update: https://github.com/w3c/web-platform-tests/pull/7347
  * [x] Make sure that events can't be recycled. 
  * [x] Make sure that events become useless after turn of event loop.

### PR DESCRIPTION
**Pull request by marcoscaceres**
Closes #7288

### CODE REVIEW

**Comment by marcoscaceres** [comment]
@rsolomakhin, just noting that Chrome is currently failing the first part of this test - I've not filed a bug yet tho.

**Comment by w3c-bots** [comment]
# Build PASSED

Started: 2017-09-18 04:01:44
Finished: 2017-09-18 04:06:49


<h2>Failing Jobs</h2>
<ul>

<li>safari:10.0</li>

<li>MicrosoftEdge:14.14393</li>

</ul>




View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/18603)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/276693946)

**Comment by rsolomakhin** [comment]
> request.[[state]] is not interactive, must throw an InvalidStateError.

This part?

**Comment by marcoscaceres** [comment]
> This part?
Yeah... it doesn't seem to throw.

File: payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html [hunk 1]
```diff
@@ -0,0 +1,93 @@
+<!doctype html>
+<meta charset="utf8">
+<link rel="help" href="https://www.w3.org/TR/payment-request/#updatewith()-method">
+<title>
+</title>
```
Comment (edenchuang) [inline payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html]: No title here.

File: payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  const request = new PaymentRequest(validMethods, validDetails, validOptions);
+  const eventPromise = new Promise(resolve => {
+    request.addEventListener("shippingaddresschange", resolve);
+  });
+  const responsePromise = request.show();
+  return { eventPromise, responsePromise };
+}
+
+function testRequestIsClosed({ textContent: testName }) {
+  promise_test(async t => {
```
Comment (edenchuang) [inline payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html]: Maybe disable the passed in button.

File: payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        event.updateWith(validDetails);
+      },
+      "request.[[state]] is not interactive, must throw an InvalidStateError."
+    );
+    responsePromise.catch(err => err);
+  }, testName.trim());
+}
+
+function testRequestIsUpdating({ textContent: testName }) {
+  promise_test(async t => {
```
Comment (edenchuang) [inline payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html]: Might disable the passed in button.

File: payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    const { eventPromise, responsePromise } = getPaymentPromises();
+    const event = await eventPromise;
+    // Set request.[[updating]] to true
+    event.updateWith(validDetails);
+    assert_throws(
+      "InvalidStateError",
+      () => {
+        event.updateWith(validDetails);
```
Comment (edenchuang) [inline payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html]: I guess there is probably a race condition.
We can't make sure event.updateWith() is called before the internal handling of the last updateWith() finishes.
I suggest that using async function to create a promise and don't resolve/reject it to make sure the that PaymentRequest.[[state]] is updating.

async function detailsPromise() {
    return new Promise((resolved, rejected) => {});
}
event.updateWith(detailsPromise);
assert_throws(
    ...
    event.updateWith(validDetails);
    ...
}

**Review by edenchuang — CHANGES_REQUESTED** [review]
Please fix the potential race condition issue.

File: payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html (hunk 4 above)
Comment (marcoscaceres) [reply payment-request/PaymentRequestUpdateEvent/updateWith-state-checks-manual.https.html]: Good catch!

**Comment by marcoscaceres** [comment]
@edenchuang, wdyt?

**Review by edenchuang — APPROVED** [review]

**Pull request merged**